content/posts/dynamic-email-routing-with-mailgun-and-python.md

```diff
@@ -0,0 +1,209 @@
+---
+title: "Dynamic Email Routing With Mailgun and Python"
+date: 2019-10-26T10:15:52-03:00
+draft: false
+authors: ['mauricio']
+tags: ['automation', 'mailgun', 'SMTP', 'custom domain']
+---
+# First things first
+
+This is a three part article. The index below will be updated with the respective links as soon as the upcoming parts are completed an published.
+
+1. [Creating the Base Function and Receiving POSTs](.)
+2. Automated Hosting of the Cloud Function
+3. Routing Received Emails
+
+# What do we want to achieve?
+
+[destroy.systems](/) is not a huge company. It's not even a company. Actually, it's just [some folks](/authors) who want to write about technical stuff and share interesting experiments. Not being a huge company also means that we don't have lots of resources to spare, we bought a domain and that's already a lot, considering that it won't generate any revenue. Having the domain, we would love to have some custom emails but, for sure, on the cheap side. We would like to have addresses for services as well (github@destroy.systems, circleci@destroy.systems, for example) and have this automatically routed to somewhere else.
+
+Then you'll say: _you can actually do that, there's this fancy tool called [Mailgun](https://www.mailgun.com) that probably achieves what you folks want._
+
+True. We use it, and it's awesome. But stay with me for a second.
+
+Let's say that my email is personalemail@gmail.com, to which we can [add some extra information](https://thenextweb.com/google/2017/08/17/how-the-plus-sign-can-save-your-gmail-inbox-from-becoming-a-pit-of-doom/) in the address, by adding _+something_ after the username. So, if we wanted to track our emails from Github, for example, we can subscribe to github with personalemail+github@gmail.com, and with that we can filter our emails by tracking those that were sent to that address, instead of my default one.
+
+This is such a great feature that Gmail offers. Now imagine if I wanted to do something similar _with my own domain_. Or, even better, create addresses __on the fly__. I'd love to have an address nameofservice@destroy.systems, and then redirect them to personalemail+nameofservice@gmail.com. This is easily achievable by creating new rules on mailgun, the problem is that we would need to create a new rule for every new service. And we couldn't even have a generic case, as the receiver's address cannot be parametrized, everything would have to go to the same address. We need to use something else.
+
+# What are the tools available?
+
+Mailgun has an API. By using this and some scripting we can achieve what we want. Then, if we host this script somewhere, we can have it permanently accessible. Think about _cloud_, _serverless_ and all of these fancy words. We'll talk about in in part 2.
```

Comment:
`in in` ---> `that in`

Reply:
Fix'd here https://github.com/destroysystems/src/pull/5/commits/ac1a68de897f52b4175b166f51875734410f78b9